src/Rings.js

```diff
@@ -0,0 +1,728 @@
+
+    const secondaries = [];
+    let total = 0;
+
+    primaries.forEach(p => {
+      const primaryId = p.id;
+
+      p.edges = p.edges.filter(edge => !claimed.includes(edge.source.id) && edge.target.id === primaryId ||
+                                       !claimed.includes(edge.target.id) && edge.source.id === primaryId);
+
+      total += p.edges.length || 1;
+
+      p.edges.forEach(edge => {
+        const {source, target} = edge;
+        const claim = target.id === primaryId ? source : target;
+        claimed.push(claim.id);
+      });
+    });
+
+    const radian = Math.PI * 2;
+    let offset = 0;
+
+    primaries.forEach((p, i) => {
+      const children = p.edges.length || 1;
+      const space = radian / total * children;
+
+      if (i === 0) {
+        offset -= space / 2;
+      }
+
+      let angle = offset + space / 2;
+
+      angle -= radian / 4;
+
+      p.radians = angle;
+      p.x = width / 2 + primaryRing * Math.cos(angle);
+      p.y = height / 2 + primaryRing * Math.sin(angle);
+
+      offset += space;
+
+      p.edges.forEach((edge, i) => {
+        const node = edge.source.id === p.id ? edge.target : edge.source;
+        const s = radian / total;
+        const a = angle - s * children / 2 + s / 2 + s * i;
+
+        node.radians = a;
+        node.x = width / 2 + secondaryRing * Math.cos(a);
+        node.y = height / 2 + secondaryRing * Math.sin(a);
+        
+        secondaries.push(node);
+      });
+    });
+
+    const primaryDistance = ringWidth / 2;
+    const secondaryDistance = ringWidth / 4;
+
+    let primaryMax = primaryDistance / 2 - 4;
+    if (primaryDistance / 2 - 4 < 8) {
+      primaryMax = min([primaryDistance / 2, 8]);
+    }
+
+    let secondaryMax = secondaryDistance / 2 - 4;
+    if (secondaryDistance / 2 - 4 < 4) {
+      secondaryMax = min([secondaryDistance / 2, 4]);
+    }
+
+    if (secondaryMax > ringWidth / 10) {
+      secondaryMax = ringWidth / 10;
+    }
+
+    if (secondaryMax > primaryMax && secondaryMax > 10) {
+      secondaryMax = primaryMax * .75;
+    }
+    if (primaryMax > secondaryMax * 1.5) {
+      primaryMax = secondaryMax * 1.5;
+    }
+
+    primaryMax = Math.floor(primaryMax);
+    secondaryMax = Math.floor(secondaryMax);
+
+    let radiusFn;
+
+    if (this._size) {
+      const domain = extent(data, d => d.size);
+
+      if (domain[0] === domain[1]) {
+        domain[0] = 0;
+      }
+
+      radiusFn = scales.scaleLinear()
+        .domain(domain)
+        .rangeRound([3, min([primaryMax, secondaryMax])]);
+
+      const val = center.size;
+      center.r = radiusFn(val);
+    }
+    else {
+      radiusFn = scales.scaleLinear()
+        .domain([1, 2])
+        .rangeRound([primaryMax, secondaryMax]);
+    }
+
+    secondaries.forEach(s => {
+      s.ring = 2;
+      const val = this._size ? s.size : 2;
+      s.r = this._sizeMin ? max([this._sizeMin, radiusFn(val)]) : this._sizeMax ? min([this._sizeMax, radiusFn(val)]) : radiusFn(val);
```

Comment:
both the primary and secondary forEach loops here should already have the site pre-calculated from the top of the file, so you should just be able to pass `s.r` and `p.r` through the radius function?

Reply:
The `r` values calculated at the top of the file are incorrect. I am going to remove them and keep the calculations here in the forEach loops instead.